sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceNullWithFalseInPredicate.scala

```diff
@@ -94,6 +94,7 @@ object ReplaceNullWithFalseInPredicate extends Rule[LogicalPlan] {
         replaceNullWithFalse(cond) -> replaceNullWithFalse(value)
       }
       val newElseValue = cw.elseValue.map(replaceNullWithFalse)
+        .orElse(if (newBranches.forall(_._2 == FalseLiteral)) Some(FalseLiteral) else None)
```

Comment:
This change assumes that other rules can optimize CASE WHEN with false in all branches, which violates the catalyst principle that rules should be orthogonal. Can we think of a better way?

Reply:
Here we do need the logic from 2 optimizer rules:
1. null means false in predicates
2. eliminate CASE WHEN if all branches return the same value

We can probably include logic 2 in this rule
```
...
if (newBranches.forall(_._2 == FalseLiteral) && cw.elseValue.isEmpty) {
  FalseLiteral
} else {
  val newElseValue = cw.elseValue.map(replaceNullWithFalse)
  CaseWhen(newBranches, newElseValue)
}
```

Reply:
Yes. This change assumes CaseWhen has optimized by `PushFoldableIntoBranches`, `SimplifyConditionals` and `ConstantFolding`.